cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,328 @@
+/**
+ * @file flow_layout.cpp
+ * @brief Implementation of the FlowLayout class, a custom layout for dynamically organizing widgets
+ * in rows within the constraints of available width or parent scroll areas.
+ */
+
+#include "flow_layout.h"
+
+#include "../widgets/general/layout_containers/flow_widget.h"
+
+#include <QDebug>
+#include <QLayoutItem>
+#include <QScrollArea>
+#include <QStyle>
+
+/**
+ * @brief Constructs a FlowLayout instance with the specified parent widget, margin, and spacing values.
+ * @param parent The parent widget for this layout.
+ * @param margin The layout margin.
+ * @param hSpacing The horizontal spacing between items.
+ * @param vSpacing The vertical spacing between items.
+ */
+FlowLayout::FlowLayout(QWidget *parent, int margin, int hSpacing, int vSpacing)
+    : QLayout(parent), horizontalMargin(hSpacing), verticalMargin(vSpacing)
+{
+    setContentsMargins(margin, margin, margin, margin);
+}
+
+/**
+ * @brief Destructor for FlowLayout, which cleans up all items in the layout.
+ */
+FlowLayout::~FlowLayout()
+{
+    QLayoutItem *item;
+    while ((item = takeAt(0))) {
+        delete item;
+    }
+}
+
+/**
+ * @brief Indicates the layout's support for expansion in both horizontal and vertical directions.
+ * @return The orientations (Qt::Horizontal | Qt::Vertical) this layout can expand to fill.
+ */
+Qt::Orientations FlowLayout::expandingDirections() const
+{
+    return Qt::Horizontal | Qt::Vertical;
+}
+
+/**
+ * @brief Indicates that this layout's height depends on its width.
+ * @return True, as the layout adjusts its height to fit the specified width.
+ */
+bool FlowLayout::hasHeightForWidth() const
+{
+    return true;
+}
+
+/**
+ * @brief Calculates the required height to display all items within the specified width.
+ * @param width The available width for arranging items.
+ * @return The total height needed to fit all items in rows constrained by the specified width.
+ */
+int FlowLayout::heightForWidth(int width) const
+{
+    int height = 0;
+    int rowWidth = 0;
+    int rowHeight = 0;
+
+    for (QLayoutItem *item : items) {
+        if (!(item == nullptr || item->isEmpty())) {
+            int itemWidth = item->sizeHint().width() + horizontalSpacing();
+            if (rowWidth + itemWidth > width) { // Start a new row if the row width exceeds available width
+                height += rowHeight + verticalSpacing();
+                rowWidth = itemWidth;
```

Comment:
rowWidth doesn't seem to be used anywhere?

Reply:
rowWidth is used in the if clause on Line 72.